testdata/environment_test.txt
Test LoadEnv.

<<< environment.json
{
  "jobs": {
    "Scheduler hourly 5m": {
      "path": "/scheduler-hourly-5m",
      "commands": [
        "echo Test job scheduler hourly per 5m"
      ],
      "last_run": "0001-01-01T00:00:00Z",
      "next_run": "0001-01-01T00:00:00Z",
      "id": "",
      "name": "",
      "schedule": "hourly@0,5,10,15,20,25,30,35,40,45,50,55"
    },
    "Scheduler minutely": {
      "path": "/scheduler-minutely",
      "commands": [
        "echo Test job scheduler per minute"
      ],
      "last_run": "0001-01-01T00:00:00Z",
      "next_run": "0001-01-01T00:00:00Z",
      "id": "",
      "name": "",
      "schedule": "minutely"
    },
    "Test auth_kind github": {
      "path": "/github",
      "auth_kind": "github",
      "commands": [
        "echo auth_kind is github"
      ],
      "last_run": "0001-01-01T00:00:00Z",
      "next_run": "0001-01-01T00:00:00Z",
      "id": "",
      "name": ""
    },
    "Test interval 1m": {
      "commands": [
        "echo Test job interval 1m",
        "echo Counter is $KARAJO_JOB_COUNTER"
      ],
      "last_run": "0001-01-01T00:00:00Z",
      "next_run": "0001-01-01T00:00:00Z",
      "id": "",
      "name": "",
      "interval": 60000000000
    },
    "Test long running": {
      "path": "/test-long-running",
      "commands": [
        "for ((x=0; x\u003c90; x++)); do echo \"$x\"; sleep 1; done"
      ],
      "last_run": "0001-01-01T00:00:00Z",
      "next_run": "0001-01-01T00:00:00Z",
      "id": "",
      "name": "",
      "description": "The job to test log refresh."
    },
    "Test manual run": {
      "path": "/test-manual-run",
      "commands": [
        "echo Test job manual run",
        "echo Counter is $KARAJO_JOB_COUNTER"
      ],
      "last_run": "0001-01-01T00:00:00Z",
      "next_run": "0001-01-01T00:00:00Z",
      "id": "",
      "name": "",
      "description": "The job to test manual run."
    },
    "notif-email-success": {
      "path": "/notif-email-success",
      "commands": [
        "echo Test email notification"
      ],
      "last_run": "0001-01-01T00:00:00Z",
      "next_run": "0001-01-01T00:00:00Z",
      "id": "",
      "name": "",
      "description": "Send notification when job success.",
      "notif_on_success": [
        "email-to-shulhan",
        "email-to-ops"
      ],
      "notif_on_failed": [
        "email-to-shulhan"
      ]
    },
    "test fail": {
      "path": "/test-fail",
      "auth_kind": "hmac-sha256",
      "commands": [
        "echo Test job fail",
        "echo Counter is $KARAJO_JOB_COUNTER",
        "x=$(($RANDOM%10)) \u0026\u0026 echo sleep in ${x}s \u0026\u0026 sleep $x",
        "command-not-found"
      ],
      "last_run": "0001-01-01T00:00:00Z",
      "next_run": "0001-01-01T00:00:00Z",
      "id": "",
      "name": ""
    },
    "test random": {
      "path": "/test-random",
      "commands": [
        "echo Test job random",
        "echo Counter is $KARAJO_JOB_COUNTER",
        "x=$(($RANDOM%10)) \u0026\u0026 echo sleep in ${x}s \u0026\u0026 sleep $x",
        "rand=$(($RANDOM%2)) \u0026\u0026 echo $rand \u0026\u0026 exit $rand"
      ],
      "last_run": "0001-01-01T00:00:00Z",
      "next_run": "0001-01-01T00:00:00Z",
      "id": "",
      "name": "",
      "description": "Test running command with random exit status"
    },
    "test success": {
      "path": "/test-success",
      "commands": [
        "echo Test job success",
        "echo Counter is $KARAJO_JOB_COUNTER",
        "x=$(($RANDOM%10)) \u0026\u0026 echo sleep in ${x}s \u0026\u0026 sleep $x"
      ],
      "last_run": "0001-01-01T00:00:00Z",
      "next_run": "0001-01-01T00:00:00Z",
      "id": "",
      "name": ""
    }
  },
  "http_jobs": {
    "Scheduler hourly 10m": {
      "http_method": "POST",
      "http_url": "/karajo/api/job_exec/run/test-manual-run",
      "http_request_type": "json",
      "last_run": "0001-01-01T00:00:00Z",
      "next_run": "0001-01-01T00:00:00Z",
      "id": "",
      "name": "",
      "description": "Test triggering job using scheduler hourly per 10 minutes",
      "schedule": "hourly@0,10,20,30,40,50",
      "http_timeout": 0
    },
    "Test fail": {
      "http_method": "POST",
      "http_url": "http://127.0.0.1:31937/karajo/api/job_exec/run/test-fail",
      "http_request_type": "json",
      "http_headers": [
        "A: B",
        "C: D"
      ],
      "last_run": "0001-01-01T00:00:00Z",
      "next_run": "0001-01-01T00:00:00Z",
      "id": "",
      "name": "",
      "description": "The job to test what the user interface and logs look likes if its \u003cb\u003efail\u003c/b\u003e.",
      "interval": 20000000000,
      "http_timeout": 0
    },
    "Test long running": {
      "http_method": "POST",
      "http_url": "/karajo/api/job_exec/run/test-long-running",
      "http_request_type": "json",
      "last_run": "0001-01-01T00:00:00Z",
      "next_run": "0001-01-01T00:00:00Z",
      "id": "",
      "name": "",
      "description": "The job to test job log refresh.",
      "interval": 120000000000,
      "http_timeout": 0
    },
    "Test random": {
      "http_method": "POST",
      "http_url": "/karajo/api/job_exec/run/test-random",
      "http_request_type": "json",
      "last_run": "0001-01-01T00:00:00Z",
      "next_run": "0001-01-01T00:00:00Z",
      "id": "",
      "name": "",
      "description": "Test triggering job /test-random",
      "http_timeout": 0
    },
    "Test success": {
      "http_method": "POST",
      "http_url": "/karajo/api/job_exec/run/test-success",
      "http_request_type": "json",
      "http_headers": [
        "X: Y"
      ],
      "last_run": "0001-01-01T00:00:00Z",
      "next_run": "0001-01-01T00:00:00Z",
      "id": "",
      "name": "",
      "description": "The job to test what the user interface and logs look likes if its \u003ci\u003esuccess\u003c/i\u003e.",
      "interval": 20000000000,
      "http_timeout": 0
    }
  },
  "name": "My karajo",
  "listen_address": "127.0.0.1:31937",
  "dir_base": "testdata",
  "dir_public": "testdata",
  "version": "0.987",
  "http_timeout": 300000000000,
  "max_job_running": 2,
  "is_development": false
}


CHANGELOG.adoc
= karajo changelog
:toc:
:sectanchors:
:sectlinks:


[#v0_9_2]
== karajo v0.9.2 (2024-09-08)

[#v0_9_2__bug_fix]
=== Bug fix

make: fix file permissions when installing public index.html::
+
Using 0640 (where user and group owner set to root) cause karajo
service--that run as karajo user--unable to read the file.


[#v0_9_1]
== karajo v0.9.1 (2024-04-06)

This release mostly contains chores.

env: remove [rand.Seed] usage::
+
The [ascii.Random] generate random using "crypto/rand", so no need to
seed it anymore.

all: replace module "share" with "pakakeh.go"::
+
The "share" module repository has been moved to SourceHut, with new name
"pakakeh.go".


[#v0_9_0]
== karajo v0.9.0 (2024-02-08)

[#v0_9_0__breaking_changes]
=== Breaking changes

all: refactoring JobExec APIs to have "_exec" suffix::
+
In JobHttp, we have "_http" suffix for its HTTP APIs.
To make it consistent we changes the HTTP API path to have "_exec" suffix.


all: apply default revive suggestions::
+
--
I prefer zero configuration rather that creating exclusions, like
"revive.toml" file that we have earlier, even thought it will cause
breaking changes to our APIs.

Breaking changes,

* [Env.HttpJobs] become [Env.HTTPJobs]
* [Env.HttpTimeout] become [Env.HTTPTimeout]
* [Env.HttpJobs] become [Env.HTTPJobs]
* and many more.
--

[#v0_9_0__new_features]
=== New features

all: implement API to cancel running job::
+
--
In the JobBase we add method Cancel to cancel running JobExec or JobHTTP.

In the HTTP server, we add endpoint "POST /karajo/api/job_exec/cancel"
to cancel JobExec by its ID.

Implements: https://todo.sr.ht/~shulhan/karajo/1
--


[#v0_9_0__enhancements]
=== Enhancements

all: export the HTTP server field in Karajo::
+
By exporting the HTTP server field, user of Karajo can register additional
HTTP endpoints without creating new HTTP server instance.


[#v0_9_0__bug_fixes]
=== Bug fixes

all: always call finish even if the job is paused::
+
--
This is to make the [JobBase.NextRun] always set to next interval or
schedule.

Fixes: https://todo.sr.ht/~shulhan/karajo/2
--


_sys: set systemd unit to start after network.target::
+
This is to fix karajo failed to start because the DNS has not working
yet when initializing email notification.


[#v0_8_0]
== karajo v0.8.0 (2023-11-10)

[#v0_8_0__new_features]
=== New features

all: implement notification using email::
+
Karajo server now support sending notification when the job success or
failed with inline log inside the email body.

[#v0_8_0__breaking_changes]
=== Breaking changes

all: change the JobHttp log to use the same as Job::
+
--
Previously, the JobHttp use single file for log with limited size.
The way it works is quite complex, we need to maintain one file and a
buffer that able to truncate the log if its reached its max size.
We also need to store the job state in separate file.

In this changes, we replace the JobHttp log mechanism to use the same
as Job, where each execution will be log separately.
Not only this give us less complex code, it also remove some duplicate
code.

In the HTTP API we changes the path to get the JobHttp log to match with
Job path.
--

[#v0_8_0__bug_fixes]
=== Bug fixes

all: fix the HTTP API to get the Job log::
+
The loop set job to non-nil if the Job ID not found, which may return
the wrong Job log.

all: changes the HTTP response code for a success JobExec run to 200::
+
--
Previously, in the HTTP endpoint for running a job, we return HTTP status
code 202 (Accepted) on success, but in the WUI we check using 200 (OK).

This changes fix this by return HTTP status code 200 to make it consistent
with other endpoints.
--

[#v0_8_0__enhancements]
=== Enhancements

www/karajo: changes the right status::
+
For Job with interval or schedule based show the Next run counter in
hours, minutes, and seconds.
Other Job type (WebHook) will be display the Last time its executed.

job_log: store the log content under unexported field content::
+
--
The goal is to minimize response size in API get environment.
Imagine we have 10 jobs with 10 logs each, where each log may contains
~10KB in size, in total we will return 10*10*10 KB in response body.

To minimze this, return the log content only when requested through
API to get Job or JobHttp Log.
--

karajo/app: do not auto refresh the dashboard::
+
Adding auto refresh consume more resources (bandwidth, cpu) on both
side, client and server.
The job logs or information is rarely changes, so no need to auto refresh
it.

cmd/karajo: set default configuration to "/etc/karajo/karajo.conf"::


[#v0_7_0]
== karajo v0.7.0 (2023-05-10)

This release add login feature to Karajo using user name and password that are
pre-defined in the user.conf.

[#v0_7_0__breaking_changes]
===  Breaking changes

all: remove MaxRunning and NumRunning from JobBase::
+
A job should be only run once at a time.
If we allow the same job run more than once at the same time, there
would be race condition in the command or Call that need to be handled
by user.

[#v0_7_0__new_features]
===  New features

all: implement login page::
+
--
The karajo status page now moved to /karajo/app/, while the old /karajo/
page is used for login.

The login page will be display only if Environment.Users is not empty,
otherwise user will be redirected to app page automatically.
--

[#v0_7_0__bug_fixes]
===  Bug fixes

all: fix possible lock on API environment::
+
--
Sometimes the request to /karajo/api/environment does not return any
result.  The only explanation is something lock the resource so we cannot
lock it and it will wait forever.
--

[#v0_7_0__enhancements]
===  Enhancements

all: changes on how the job queued using channel::
+
--
Previously, a job run using the following flow:

* interval/scheduler timer kicking in
* send the job finish to finished channel

If the job triggered from HTTP request, it will run on its own goroutine.

This changes add third channel, startq, to JobBase that queue the Job.

env.go
// SPDX-FileCopyrightText: 2021 M. Shulhan <ms@kilabit.info>
// SPDX-License-Identifier: GPL-3.0-or-later

package karajo

import (
	"errors"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"strings"
	"time"

	"git.sr.ht/~shulhan/pakakeh.go/lib/ascii"
	libhtml "git.sr.ht/~shulhan/pakakeh.go/lib/html"
	"git.sr.ht/~shulhan/pakakeh.go/lib/ini"
	"git.sr.ht/~shulhan/pakakeh.go/lib/mlog"
)

const (
	defDirBase       = `/`
	defEnvName       = `karajo`
	defHTTPTimeout   = 5 * time.Minute
	defListenAddress = `127.0.0.1:31937`
	defMaxJobRunning = 1
)

// Env contains configuration for HTTP server, logs, and list of jobs.
type Env struct {
	// List of JobExec by name.
	ExecJobs map[string]*JobExec `ini:"job" json:"jobs"`

	// List of JobHTTP by name.
	HTTPJobs map[string]*JobHTTP `ini:"job.http" json:"http_jobs"`

	// Notif contains list of notification setting.
	Notif map[string]EnvNotif `ini:"notif" json:"-"`

	// Index of notification client by its name.
	notif map[string]notifClient

	// Users list of user that can access web user interface.
	// The list of user optionally loaded from
	// $DirBase/etc/karajo/user.conf if the file exist.
	Users map[string]*User `json:"-"`

	// Name of the service.
	// The Name will be used for title on the web user interface, as log
	// prefix, as file prefix on the jobs state, and as file prefix on
	// log files.
	// If this value is empty, it will be set to "karajo".
	Name string `ini:"karajo::name" json:"name"`
	name string

	// Define the address for WUI, default to ":31937".
	ListenAddress string `ini:"karajo::listen_address" json:"listen_address"`

	// DirBase define the base directory where configuration, job state,
	// and job log stored.
	// This field is optional, default to current directory.
	// The structure of directory follow the common UNIX system,
	//
	//	$DirBase
	//	|
	//	+-- /etc/karajo/ +-- karajo.conf
	//	|                +-- job.d/
	//	|                +-- job_http.d/
	//	|
	//	+-- /var/lib/karajo/ +-- job/$JobExec.ID
	//	|                    +-- job_http/$JobHTTP.ID
	//	|
	//	+-- /var/log/karajo/ +-- job/$JobExec.ID
	//	|                    +-- job_http/$JobHTTP.ID
	//	|
	//	+-- /var/run/karajo/job_http/$JobHTTP.ID
	//
	// Each job log stored under directory /var/log/karajo/job and the job
	// state under directory /var/run/karajo/job.
	DirBase string `ini:"karajo::dir_base" json:"dir_base"`

	// Equal to $DirBase/etc/karajo/
	dirConfig string

	// dirConfigJobd is the directory where job configuration loaded.
	// This is to simplify managing job by splitting it per file.
	// Each job configuration end with `.conf`.
	dirConfigJobd string

	// dirConfigJobHTTPd the directory where JobHTTP configuration
	// loaded.
	// This is to simplify managing JobHTTP by splitting it per file.
	// Each JobHTTP configuration end with `.conf`.
	dirConfigJobHTTPd string

	dirLibJob     string
	dirLibJobHTTP string

	dirLogJob     string
	dirLogJobHTTP string

	// dirRunJobHTTP define the directory where JobHTTP state is stored.
	dirRunJobHTTP string

	file string

	// DirPublic define a path to serve to public.
	// While the WUI is served under "/karajo", a directory DirPublic
	// will be served under "/".
	// A DirPublic can contains sub directory as long as its name is not
	// "karajo".
	DirPublic string `ini:"karajo::dir_public" json:"dir_public"`

	Version string `json:"version"`

	// Secret define the default secret to authorize the incoming HTTP
	// request.
	// The signature is generated from HTTP payload (query or body) with
	// HMAC+SHA-256.
	// The signature is read from HTTP header "X-Karajo-Sign" as hex
	// string.
	// This field is optional, if its empty the new secret will be
	// generated and printed to standard output on each run.
	Secret  string `ini:"karajo::secret" json:"-"`
	secretb []byte

	// HTTPTimeout define the global HTTP client timeout when executing
	// each jobs.
	// This field is optional, default to 5 minutes.
	// The value of this option is using the Go [time.Duration]
	// format, for example, "30s" for 30 seconds, "1m" for 1 minute.
	HTTPTimeout time.Duration `ini:"karajo::http_timeout" json:"http_timeout"`

	// MaxJobRunning define the maximum job running at the same time.
	// This field is optional default to 1.
	MaxJobRunning int `ini:"karajo::max_job_running" json:"max_job_running"`

	// IsDevelopment if its true, the files in DirPublic will be loaded
	// directly from disk instead from embedded memfs.
	IsDevelopment bool `ini:"karajo::is_development" json:"is_development"`
}

// LoadEnv load the configuration from the ini file format.
func LoadEnv(file string) (env *Env, err error) {
	var (
		logp = `LoadEnv`
		cfg  *ini.Ini
	)

	cfg, err = ini.Open(file)
	if err != nil {
		return nil, fmt.Errorf(`%s: %w`, logp, err)
	}

	env = &Env{
		file:    file,
		Version: Version,
	}

	err = cfg.Unmarshal(env)
	if err != nil {
		return nil, fmt.Errorf(`%s: %w`, logp, err)
	}

	return env, nil
}

// NewEnv create and initialize new Env with default values,
// where Name is "karajo", listen address is ":31937", base directory is "/",
// HTTP timeout is 5 minutes, and maximum job running is 1.
func NewEnv() (env *Env) {
	env = &Env{
		Name:          defEnvName,
		ExecJobs:      make(map[string]*JobExec),
		HTTPJobs:      make(map[string]*JobHTTP),
		Users:         make(map[string]*User),
		ListenAddress: defListenAddress,
		DirBase:       defDirBase,
		Version:       Version,
		HTTPTimeout:   defHTTPTimeout,
		MaxJobRunning: defMaxJobRunning,
	}
	return env
}

// ParseEnv parse the environment from raw bytes.
func ParseEnv(content []byte) (env *Env, err error) {
	var (
		logp = `ParseEnv`
	)

	env = &Env{
		Version: Version,
	}

	err = ini.Unmarshal(content, env)
	if err != nil {
		return nil, fmt.Errorf(`%s: %w`, logp, err)
	}

	return env, nil
}

// jobExec get the JobExec by its ID.
func (env *Env) jobExec(id string) (job *JobExec) {
	for _, job = range env.ExecJobs {
		if job.ID == id {
			return job
		}
	}
	return nil
}

// jobHTTP get the registered JobHTTP by its ID.
func (env *Env) jobHTTP(id string) (job *JobHTTP) {
	for _, job = range env.HTTPJobs {
		if job.ID == id {
			return job
		}
	}
	return nil
}

func (env *Env) init() (err error) {
	var (
		logp = `init`

		job     *JobExec
		jobHTTP *JobHTTP
		name    string
	)

	if len(env.Name) == 0 {
		env.Name = defEnvName
	}
	env.name = libhtml.NormalizeForID(env.Name)

	if len(env.ListenAddress) == 0 {
		env.ListenAddress = defListenAddress
	}
	if env.HTTPTimeout == 0 {
		env.HTTPTimeout = defHTTPTimeout
	}
	if env.MaxJobRunning <= 0 {
		env.MaxJobRunning = defMaxJobRunning
	}

	if len(env.Secret) == 0 {
		var secret = ascii.Random([]byte(ascii.LettersNumber), 32)
		env.Secret = string(secret)

		mlog.Outf(`!!! WARNING: Your secret is empty and has been generated: %s`, secret)
	}
	env.secretb = []byte(env.Secret)

	err = env.initDirs()
	if err != nil {
		return fmt.Errorf(`%s: %w`, logp, err)
	}

	err = env.initNotifs()
	if err != nil {
		return fmt.Errorf(`%s: %w`, logp, err)
	}

	err = env.initUsers()
	if err != nil {
		return fmt.Errorf(`%s: %w`, logp, err)
	}

	err = env.loadJobd()
	if err != nil {
		return fmt.Errorf(`%s: %w`, logp, err)
	}

	for name, job = range env.ExecJobs {
		err = job.init(env, name)
		if err != nil {
			return fmt.Errorf(`%s: %w`, logp, err)
		}
	}

	err = env.loadJobHTTPd()
	if err != nil {
		return fmt.Errorf(`%s: %w`, logp, err)
	}

	for name, jobHTTP = range env.HTTPJobs {
		err = jobHTTP.init(env, name)
		if err != nil {
			return fmt.Errorf(`%s: %w`, logp, err)
		}
	}

	return nil
}

// initDirs create all job and log directories.
func (env *Env) initDirs() (err error) {
	var (
		logp = `initDirs`
	)

	if len(env.DirBase) == 0 {
		env.DirBase = defDirBase
	}

	env.dirConfig = filepath.Join(env.DirBase, `etc`, defEnvName)
	env.dirConfigJobd = filepath.Join(env.DirBase, `etc`, defEnvName, `job.d`)
	env.dirConfigJobHTTPd = filepath.Join(env.DirBase, `etc`, defEnvName, `job_http.d`)

	env.dirLibJob = filepath.Join(env.DirBase, `var`, `lib`, defEnvName, `job`)
	err = os.MkdirAll(env.dirLibJob, 0700)
	if err != nil {
		return fmt.Errorf(`%s: %s: %w`, logp, env.dirLibJob, err)
	}

	env.dirLibJobHTTP = filepath.Join(env.DirBase, `var`, `lib`, defEnvName, `job_http`)
	err = os.MkdirAll(env.dirLibJobHTTP, 0700)
	if err != nil {
		return fmt.Errorf(`%s: %s: %w`, logp, env.dirLibJobHTTP, err)
	}

	env.dirLogJob = filepath.Join(env.DirBase, `var`, `log`, defEnvName, `job`)
	err = os.MkdirAll(env.dirLogJob, 0700)
	if err != nil {
		return fmt.Errorf(`%s: %s: %w`, logp, env.dirLogJob, err)
	}

	env.dirLogJobHTTP = filepath.Join(env.DirBase, `var`, `log`, defEnvName, `job_http`)
	err = os.MkdirAll(env.dirLogJobHTTP, 0700)
	if err != nil {
		return fmt.Errorf(`%s: %s: %w`, logp, env.dirLogJobHTTP, err)
	}

	env.dirRunJobHTTP = filepath.Join(env.DirBase, `var`, `run`, defEnvName, `job_http`)
	err = os.MkdirAll(env.dirRunJobHTTP, 0700)
	if err != nil {
		return fmt.Errorf(`%s: %s: %w`, logp, env.dirRunJobHTTP, err)
	}

	return nil
}

// initNotifs initialize the notification.
func (env *Env) initNotifs() (err error) {
	var (
		logp = `initNotifs`

		name        string
		envNotif    EnvNotif
		clientNotif notifClient
	)
	env.notif = make(map[string]notifClient)
	for name, envNotif = range env.Notif {
		envNotif.Name = name

		envNotif.init()

		clientNotif, err = envNotif.createClient()
		if err != nil {
			return fmt.Errorf(`%s: %w`, logp, err)
		}
		env.notif[name] = clientNotif
	}
	return nil
}

// initUsers load users for authentication from $DirBase/etc/karajo/user.conf.
func (env *Env) initUsers() (err error) {
	var (
		logp         = `initUsers`
		fileUserConf = filepath.Join(env.dirConfig, `user.conf`)

		listUser map[string]*User
		name     string
		u        *User
	)

	listUser, err = loadUsers(fileUserConf)
	if err != nil {
		return fmt.Errorf(`%s: %w`, logp, err)
	}

	mlog.Outf(`Loaded %d users from %s.`, len(listUser), fileUserConf)

	if env.Users == nil {
		env.Users = make(map[string]*User)
	}
	for name, u = range listUser {
		env.Users[name] = u
	}

	return nil
}

// loadConfigJob load jobs configuration from file.
//
// The conf file can contains one or more jobs configuration.
func (env *Env) loadConfigJob(conf string) (jobs map[string]*JobExec, err error) {
	type jobContainer struct {
		ExecJobs map[string]*JobExec `ini:"job"`
	}

	var (
		logp = `loadConfigJob`

		cfg *ini.Ini
	)

	cfg, err = ini.Open(conf)
	if err != nil {
		return nil, fmt.Errorf(`%s: %s: %w`, logp, conf, err)
	}

	var jobc = jobContainer{}

	err = cfg.Unmarshal(&jobc)
	if err != nil {
		return nil, fmt.Errorf(`%s: %s: %w`, logp, conf, err)
	}

	jobs = jobc.ExecJobs
	jobc.ExecJobs = nil

	return jobs, nil
}

// loadConfigJobHTTP load JobHTTP configuration from file.
func (env *Env) loadConfigJobHTTP(conf string) (httpJobs map[string]*JobHTTP, err error) {
	type jobContainer struct {
		HTTPJobs map[string]*JobHTTP `ini:"job.http"`
	}

	var (
		logp = `loadConfigJobHTTP`

		cfg *ini.Ini
	)

	cfg, err = ini.Open(conf)
	if err != nil {
		return nil, fmt.Errorf(`%s: %s: %w`, logp, conf, err)
	}

	var jobc = jobContainer{}

	err = cfg.Unmarshal(&jobc)
	if err != nil {
		return nil, fmt.Errorf(`%s: %s: %w`, logp, conf, err)
	}

	httpJobs = jobc.HTTPJobs
	jobc.HTTPJobs = nil

	return httpJobs, nil
}

// loadJobd load all job configurations from a directory.
func (env *Env) loadJobd() (err error) {
	var (
		logp = `loadJobd`

		jobd    *os.File
		listde  []os.DirEntry
		de      os.DirEntry
		fm      os.FileMode
		name    string
		jobConf string
		jobs    map[string]*JobExec
		job     *JobExec
	)

	jobd, err = os.Open(env.dirConfigJobd)
	if err != nil {
		if errors.Is(err, fs.ErrNotExist) {
			return nil
		}
		return fmt.Errorf(`%s: %w`, logp, err)
	}

	listde, err = jobd.ReadDir(0)
	if err != nil {
		return fmt.Errorf(`%s: %w`, logp, err)
	}

	if env.ExecJobs == nil {
		env.ExecJobs = make(map[string]*JobExec)
	}

	for _, de = range listde {
		if de.IsDir() {
			continue
		}
		fm = de.Type()
		if !fm.IsRegular() {
			continue
		}
		name = de.Name()

		if name[0] == '.' {
			// Exclude hidden file.
			continue
		}
		if !strings.HasSuffix(name, `.conf`) {
			continue
		}

		jobConf = filepath.Join(env.dirConfigJobd, name)

		jobs, err = env.loadConfigJob(jobConf)
		if err != nil {
			return fmt.Errorf(`%s: %w`, logp, err)
		}

		for name, job = range jobs {
			env.ExecJobs[name] = job
		}
	}
	return nil
}

// loadJobHTTPd load all JobHTTP configurations from a directory.
func (env *Env) loadJobHTTPd() (err error) {
	var (
		logp = `loadJobHTTPd`

		jobd     *os.File
		listde   []os.DirEntry
		de       os.DirEntry
		fm       os.FileMode
		name     string
		fileConf string
		httpJobs map[string]*JobHTTP
		httpJob  *JobHTTP
	)

	jobd, err = os.Open(env.dirConfigJobHTTPd)
	if err != nil {
		if errors.Is(err, fs.ErrNotExist) {
			return nil
		}
		return fmt.Errorf(`%s: %w`, logp, err)
	}

	listde, err = jobd.ReadDir(0)
	if err != nil {
		return fmt.Errorf(`%s: %w`, logp, err)
	}

	if env.HTTPJobs == nil {
		env.HTTPJobs = make(map[string]*JobHTTP)
	}

	for _, de = range listde {
		if de.IsDir() {
			continue
		}
		fm = de.Type()
		if !fm.IsRegular() {
			continue
		}
		name = de.Name()

		if name[0] == '.' {
			// Exclude hidden file.
			continue
		}
		if !strings.HasSuffix(name, `.conf`) {
			continue
		}

		fileConf = filepath.Join(env.dirConfigJobHTTPd, name)

		httpJobs, err = env.loadConfigJobHTTP(fileConf)
		if err != nil {
			return fmt.Errorf(`%s: %w`, logp, err)
		}

		for name, httpJob = range httpJobs {
			env.HTTPJobs[name] = httpJob
		}
	}
	return nil
}

func (env *Env) lockAllJob() {
	var job *JobExec
	for _, job = range env.ExecJobs {
		job.Lock()
	}

	var jobHTTP *JobHTTP
	for _, jobHTTP = range env.HTTPJobs {
		jobHTTP.Lock()
	}
}

func (env *Env) unlockAllJob() {
	var job *JobExec
	for _, job = range env.ExecJobs {
		job.Unlock()
	}

	var jobHTTP *JobHTTP
	for _, jobHTTP = range env.HTTPJobs {
		jobHTTP.Unlock()
	}
}
